tool-plugins/intellij/src/main/java/io/ballerina/plugins/idea/debugger/BallerinaDAPClientConnector.java

```diff
@@ -0,0 +1,224 @@
+import org.jetbrains.annotations.NotNull;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Future;
+
+import static io.ballerina.plugins.idea.BallerinaConstants.BALLERINA_DEBUG_LAUNCHER_NAME;
+import static io.ballerina.plugins.idea.BallerinaConstants.BALLERINA_DEBUG_LAUNCHER_PATH;
+
+/**
+ * Used to communicate with debug server.
+ */
+public class BallerinaDAPClientConnector {
+
+    private static final Logger LOG = Logger.getInstance(BallerinaDAPClientConnector.class);
+
+    private BallerinaDebugProcess context;
+    private Project project;
+    private String host;
+    private int port;
+    private DAPClient debugClient;
+    private IDebugProtocolServer debugServer;
+    private DAPRequestManager requestManager;
+    private BallerinaStreamConnectionProvider streamConnectionProvider;
+    private Future<Void> launcherFuture;
+    private CompletableFuture<Capabilities> initializeFuture;
+    private Capabilities initializeResult;
+    private ConnectionState myConnectionState;
+
+    private static final int DEBUG_ADAPTOR_PORT = 4711;
+    private static final String CONFIG_SOURCEROOT = "sourceRoot";
+    private static final String CONFIG_DEBUGEE_PORT = "debuggeePort";
+
+    public BallerinaDAPClientConnector(@NotNull Project project, @NotNull String host, int port) {
+        this.project = project;
+        this.host = host;
+        this.port = port;
+        myConnectionState = ConnectionState.NOT_CONNECTED;
+    }
+
+    public DAPRequestManager getRequestManager() {
+        return requestManager;
+    }
+
+    public Project getProject() {
+        return project;
+    }
+
+    public int getPort() {
+        return port;
+    }
+
+    public BallerinaDebugProcess getContext() {
+        return context;
+    }
+
+    public void setContext(BallerinaDebugProcess context) {
+        this.context = context;
+    }
+
+    void createConnection() {
+        try {
+            myConnectionState = ConnectionState.CONNECTING;
+            streamConnectionProvider = createConnectionProvider(project);
+            if (streamConnectionProvider == null) {
+                LOG.warn("Unable to establish the socket connection provider.");
+                return;
+            }
+            streamConnectionProvider.start();
+            Pair<InputStream, OutputStream> streams = new ImmutablePair<>(streamConnectionProvider.getInputStream(),
+                    streamConnectionProvider.getOutputStream());
+            InputStream inputStream = streams.getKey();
+            OutputStream outputStream = streams.getValue();
+
+            if (inputStream == null || outputStream == null) {
+                LOG.warn("Unable to establish connection with the debug server.");
```

Comment:
When both `inputStream` and `outputStream` are NULL, can the program continue further?

Reply:
Should be returned immediately if either of the streams is NULL. :) Fixed.